src/array/cpu/rowwise_pick.h

```diff
@@ -310,101 +309,116 @@ COOMatrix CSRRowWisePerEtypePick(
+              << "etype values exceed the number of fanouts";
         }
-        if (!rowwise_etype_sorted)  // the edge type is sorted, not need to sort
-                                    // it
-          std::sort(
-              et_idx.begin(), et_idx.end(),
-              [&et](IdxType i1, IdxType i2) { return et[i1] < et[i2]; });
-        CHECK_LT(et[et_idx[len - 1]], num_etypes)
-            << "etype values exceed the number of fanouts";
-
-        IdxType cur_et = et[et_idx[0]];
-        int64_t et_offset = 0;
-        int64_t et_len = 1;
-        for (int64_t j = 0; j < len; ++j) {
-          CHECK((j + 1 == len) || (et[et_idx[j]] <= et[et_idx[j + 1]]))
-              << "Edge type is not sorted. Please sort in advance or specify "
-                 "'rowwise_etype_sorted' as false.";
-          if ((j + 1 == len) || cur_et != et[et_idx[j + 1]]) {
-            // 1 end of the current etype
-            // 2 end of the row
-            // random pick for current etype
-            if ((num_picks[cur_et] == -1) ||
-                (et_len <= num_picks[cur_et] && !replace)) {
-              // fast path, select all
-              for (int64_t k = 0; k < et_len; ++k) {
-                const IdxType eid_offset = off + et_idx[et_offset + k];
+
+        IdxType cur_et;
+        int64_t et_offset;
+        int64_t et_end = 0;
+        int64_t et_len;
+        while (et_end < len) {
+          et_offset = et_end;
+          if (!et.empty()) {
+            cur_et = et[et_idx[et_offset]];
+          } else {
+            const IdxType eid_offset = off + et_idx[et_offset];
+            const IdxType homogenized_eid = eid ? eid[eid_offset] : eid_offset;
+            auto it = std::upper_bound(
+                eid2etype_offset.begin(), eid2etype_offset.end(),
+                homogenized_eid);
+            cur_et = it - eid2etype_offset.begin() - 1;
+          }
+          // Get et_end by binary search
+          auto cur_it = et_idx.begin() + et_offset;
+          auto next_it = std::upper_bound(
+              cur_it, et_idx.end(), cur_et,
+              [&](const IdxType v, const IdxType e) {
+                const IdxType eid_offset = off + et_idx[e];
                 const IdxType homogenized_eid =
                     eid ? eid[eid_offset] : eid_offset;
                 auto it = std::upper_bound(
                     eid2etype_offset.begin(), eid2etype_offset.end(),
                     homogenized_eid);
-                const IdxType heterogenized_etype =
-                    it - eid2etype_offset.begin() - 1;
-                const IdxType heterogenized_eid =
-                    homogenized_eid - eid2etype_offset[heterogenized_etype];
-
-                if (!has_probs ||
-                    IsNullArray(prob_or_mask[heterogenized_etype])) {
-                  // No probability, select all
+                auto element_et = it - eid2etype_offset.begin() - 1;
+                return v < element_et;
+              });
```

Comment:
This is nesting two `std::upper_bound`?  I don't quite understand how the outer `std::upper_bound` works.

Reply:
The outer `upper_bound` is used to find the end offset of current etype, the inner `upper_bound` is to find the heterogenized etype by searching the `etype_offset` array with `homogeneous eid`.